src/control/mod.rs

```diff
@@ -336,30 +340,60 @@ impl<T: cio::CIO> Control<T> {
     }
 
     fn handle_lst_ev(&mut self, msg: listener::Message) {
-        debug!("Adding peer for torrent with hash {:?}!", msg.hash);
-        if let Some(tid) = self.hash_idx.get(&msg.hash).cloned() {
-            let id = msg.id;
-            let rsv = msg.rsv;
-            match peer::PeerConn::new_incoming(msg.conn, msg.reader) {
-                Ok(p) => self.add_inc_peer(tid, p, id, rsv),
+        match peer::PeerConn::new_incoming(msg.conn) {
+            Ok(pconn) => match self.cio.add_peer(pconn) {
+                Ok(pid) => {
+                    self.incoming.insert(pid);
+                }
                 Err(e) => {
-                    error!("Failed to create peer connection: {:?}", e);
+                    error!("Failed to add peer connection: {:?}", e);
                 }
-            };
-        } else {
-            let h = msg.hash;
-            error!("Couldn't add peer, torrent {} doesn't exist", hash_to_id(&h));
+            },
+            Err(e) => {
+                error!("Failed to create peer connection: {:?}", e);
+            }
+        }
+    }
+
+    fn inc_handshake(
+        &mut self,
+        pid: cio::PID,
+        ev: cio::Result<torrent::Message>,
+    ) -> Result<(), ()> {
+        match ev {
+            Ok(msg) => match msg {
+                torrent::Message::Handshake { hash, id, rsv } => {
+                    debug!("Adding peer for torrent with hash {:?}!", hash_to_id(&hash));
+                    if let Some(tid) = self.hash_idx.get(&hash).cloned() {
+                        return self.add_inc_peer(tid, pid, id, rsv);
+                    } else {
+                        error!(
+                            "Couldn't add peer, torrent {} doesn't exist",
+                            hash_to_id(&hash)
+                        );
+                    }
+                }
+                _ => unreachable!(),
```

Comment:
Can you leave a comment here noting that Peer's are instantiated trying to read a handshake so the first message must be a handshake.

Reply:
Sure thing. Done.

Reply:
Is it guaranteed that the peer can't send any other msg?

Reply:
No. But it *is* guaranteed that the Reader won't accept anything but a handshake (that's what the "instantiated trying to read a handshake" means above).

Reply:
I will change the comment to make this more clear! Thanks.